File: Sources/BaseTextStorage.m
//
//  BaseTextStorage.m
//  CanvasText
//
//  Created by Sam Soffes on 6/17/16.
//  Copyright © 2016 Canvas Labs, Inc. All rights reserved.
//

#import "BaseTextStorage.h"

@interface BaseTextStorage ()
@property (nonatomic) NSMutableAttributedString *storage;
@end

@implementation BaseTextStorage

@synthesize storage = _storage;

// MARK: - Initializers

- (instancetype)init {
	if (self = [super init]) {
		self.storage = [[NSMutableAttributedString alloc] init];
	}
	return self;
}


// MARK: - NSTextStorage

- (NSString *)string {
	return self.storage.string;
}

- (NSDictionary<NSString *,id> *)attributesAtIndex:(NSUInteger)location effectiveRange:(NSRangePointer)effectiveRange {
	return [self.storage attributesAtIndex:location effectiveRange:effectiveRange];
}

- (void)replaceCharactersInRange:(NSRange)range withString:(NSString *)aString {
	[self.storage replaceCharactersInRange:range withString:aString];
	
	NSInteger change = aString.length - range.length;
	[self edited:NSTextStorageEditedCharacters range:range changeInLength:change];
}

- (void)setAttributes:(NSDictionary<NSString *,id> *)attributes range:(NSRange)range {
	if (NSMaxRange(range) > self.length) {
		NSLog(@"WARNING: Tried to set attributes at out of bounds range %@. Length: %lu", NSStringFromRange(range), (unsigned long)self.length);
		return;
	}
	
	[self beginEditing];
	[self.storage setAttributes:attributes range:range];
	[self edited:NSTextStorageEditedAttributes range:range changeInLength:0];
	[self endEditing];
}

@end

File: Sources/BaseTextStorage.swift
//
//  BaseTextContainer.swift
//  CanvasText
//
//  Created by Sam Soffes on 2/16/16.
//  Copyright © 2016 Canvas Labs, Inc. All rights reserved.
//

#if os(OSX)
	import AppKit
#else
	import UIKit
#endif

/// Concrete text storage intended to be subclassed.
public class BaseTextStorage: NSTextStorage {

	// MARK: - Properties

	let storage = NSMutableAttributedString()


	// MARK: - NSTextStorage

	public override var string: String {
		return storage.string
	}

	public override func attributesAtIndex(location: Int, effectiveRange: NSRangePointer) -> [String : AnyObject] {
		return storage.attributesAtIndex(location, effectiveRange: effectiveRange)
	}

	public override func replaceCharactersInRange(range: NSRange, withString string: String) {
		storage.replaceCharactersInRange(range, withString: string)

		let stringLength = (string as NSString).length

		let change = stringLength - range.length
		edited(.EditedCharacters, range: range, changeInLength: change)
	}

	public override func setAttributes(attributes: [String : AnyObject]?, range: NSRange) {
		guard range.max <= length else {
			print("WARNING: Tried to set attributes at out of bounds range \(range). Length: \(length)")
			return
		}

		beginEditing()

		storage.setAttributes(attributes, range: range)

		edited(.EditedAttributes, range: range, changeInLength: 0)

		endEditing()
	}
}
